src/Shared/BuildEnvironmentHelper.cs

```diff
@@ -126,190 +93,204 @@ private static BuildEnvironment Initialize()
         }
 
-        private static BuildEnvironment TryFromFolder(string folder, bool runningTests, bool runningInVisualStudio, string visualStudioPath)
+        private static BuildEnvironment TryFromMsBuildAssembly()
         {
-            if (string.IsNullOrEmpty(folder)) return null;
+            var buildAssembly = s_getExecutingAssmblyPath();
+            if (buildAssembly == null) return null;
 
-            return (
-                    s_msBuildExeNames.Select(msbuildFileName => Path.Combine(folder, msbuildFileName))
-                    .Where(IsValidMSBuildPath)
-                    .Select(msBuildPath => new BuildEnvironment(msBuildPath, runningTests, runningInVisualStudio, visualStudioPath))
-                   ).FirstOrDefault();
-        }
+            var msBuildExe = Path.Combine(FileUtilities.GetFolderAbove(buildAssembly), "MSBuild.exe");
+            var msBuildDll = Path.Combine(FileUtilities.GetFolderAbove(buildAssembly), "MSBuild.dll");
 
-        /// <summary>
-        /// Determine whether the given path is considered to be an acceptable path to MSBuild.
-        /// </summary>
-        /// <remarks>
-        /// If we are running in an orphaned way (i.e. running from Microsoft.Build.dll in someone else's process),
-        /// that folder will not be sufficient as a build tools folder (e.g. we can't launch MSBuild.exe from that
-        /// location). At minimum, it must have MSBuild.exe and MSBuild.exe.config.
-        /// </remarks>
-        /// <param name="path">Full path to MSBuild.exe</param>
-        /// <returns>True when the path to MSBuild is valid.</returns>
-        private static bool IsValidMSBuildPath(string path)
-        {
-            bool msbuildExeExists = !string.IsNullOrEmpty(path) &&
-                s_msBuildExeNames.Any(i => i.Equals(Path.GetFileName(path), StringComparison.OrdinalIgnoreCase)) &&
-                    File.Exists(path);
-#if FEATURE_SYSTEM_CONFIGURATION
-            // If we can read toolsets out of msbuild.exe.config, we must
-            // try to do so.
-            return msbuildExeExists &&
-                   File.Exists($"{path}.config");
-#else
-            // On .NET Core, we can't read the contents of msbuild.exe.config,
-            // so it doesn't matter if it exists.
-            return msbuildExeExists;
-#endif
-        }
+            bool runningTests = CheckIfRunningTests(s_getProcessFromRunningProcess());
 
-        private static bool CheckIfRunningTests(string processNameCommandLine, string processNameCurrentProcess)
-        {
-            // First check if we're running in a known test runner.
-            if (IsProcessInList(processNameCommandLine, s_testRunners) ||
-                IsProcessInList(processNameCurrentProcess, s_testRunners))
+            // First check if we're in a VS installation
+            if (Regex.IsMatch(buildAssembly, $@".*\\MSBuild\\{CurrentToolsVersion}\\Bin\\.*", RegexOptions.IgnoreCase))
```

Comment:
The logic from here on to the end of the method duplicates almost entirely the code in TryFromMsBuildProcess. Is it worth unifying?